File: gui-client/world_step/world_step/data/zone_list/zone/region/portals/portal/to/to.cs
using System;
using System.Collections.Immutable;
using System.Collections.Generic;
using System.Xml;
using System.Linq;
using Guiclient.util;
using Godot;
using XSD;

namespace XSD.Nworld_step.Ndata.Nzone_list.Nzone.Nregion.Nportals.Nportal.Nto {}
namespace XSD {
}
namespace XSD.Nworld_step.Ndata.Nzone_list.Nzone.Nregion.Nportals.Nportal {
  public class to : IEquatable<to>, XSD.ILinkedNode  {

    public static string ClassTypeId = ".world_step.data.zone_list.zone.region.portals.portal.to";
    public static string TagName = "to";

    public string NodeName {get =>"to";}
    public RawNode rawNode = new RawNode();

    private ILinkedNode? _parentNode;
    public ILinkedNode? ParentNode {get => _parentNode; set => _parentNode = value;}
    private List<Action<to>> _onSelfChangeCallbackList = new();
    private List<Action<List<ILinkedNode>>> _onChangeCallbackList = new();

    //Attributes
    private System.String _zone_ref;
    public System.String zone_ref { get => _zone_ref; set => _zone_ref = value; }
    private System.String _region_ref;
    public System.String region_ref { get => _region_ref; set => _region_ref = value; }
    private System.String _side;
    public System.String side { get => _side; set => _side = value; }
    private System.Int32 _start;
    public System.Int32 start { get => _start; set => _start = value; }
    private System.Int32 _end;
    public System.Int32 end { get => _end; set => _end = value; }

    //Children elements
    public to()
    {
    }

    public to(RawNode rawNode)
    {
      Deserialize(rawNode);
    }

    public to(XmlElement xmlElement)
    {
      this.rawNode.Deserialize(xmlElement);
      Deserialize(rawNode);
    }

    public void SetAttribute(string name, string? value)
    {
      if(name == "zone_ref")
      {
        Set_zone_ref(value);
      }
      if(name == "region_ref")
      {
        Set_region_ref(value);
      }
      if(name == "side")
      {
        Set_side(value);
      }
      if(name == "start")
      {
        Set_start(value?.ToInt() ?? 0);
      }
      if(name == "end")
      {
        Set_end(value?.ToInt() ?? 0);
      }
    }

    public void SetChild(dynamic linkedNode)
    {
    }

    public void ClearChild(dynamic linkedNode)
    {
    }

    public Action OnSelfChange(Action<to> callback)
    {
      _onSelfChangeCallbackList.Add(callback);
      return () => _onSelfChangeCallbackList.Remove(callback);
    }

    public Action OnSelfChangeNode(Action<ILinkedNode> callback)
    {
      _onSelfChangeCallbackList.Add(callback);
      return () => _onSelfChangeCallbackList.Remove(callback);
    }


    public Action OnChange(Action<List<ILinkedNode>> callback)
    {
      _onChangeCallbackList.Add(callback);
      return () => _onChangeCallbackList.Remove(callback);
    }

    public void Deserialize (RawNode rawNode)
    {
      this.rawNode = rawNode;
      // Godot.GD.Print("Deserializing to");
      //Deserialize arguments
      if(rawNode.attributes.ContainsKey("zone_ref"))
      {
        var attribute_zone_ref = rawNode.attributes["zone_ref"];
        this.zone_ref = rawNode.attributes["zone_ref"];
      }
      if(rawNode.attributes.ContainsKey("region_ref"))
      {
        var attribute_region_ref = rawNode.attributes["region_ref"];
        this.region_ref = rawNode.attributes["region_ref"];
      }
      if(rawNode.attributes.ContainsKey("side"))
      {
        var attribute_side = rawNode.attributes["side"];
        this.side = rawNode.attributes["side"];
      }
      if(rawNode.attributes.ContainsKey("start"))
      {
        var attribute_start = rawNode.attributes["start"];
        this.start = rawNode.attributes["start"];
      }
      if(rawNode.attributes.ContainsKey("end"))
      {
        var attribute_end = rawNode.attributes["end"];
        this.end = rawNode.attributes["end"];
      }

      //Deserialize children
      NotifyChange();
    }

    public RawNode SerializeIntoRawNode()
    {
      //Serialize arguments
      if(this._zone_ref != null)
      {
        rawNode.attributes["zone_ref"] = this._zone_ref.ToString();
      }
      if(this._region_ref != null)
      {
        rawNode.attributes["region_ref"] = this._region_ref.ToString();
      }
      if(this._side != null)
      {
        rawNode.attributes["side"] = this._side.ToString();
      }
      if(this._start != null)
      {
        rawNode.attributes["start"] = this._start.ToString();
      }
      if(this._end != null)
      {
        rawNode.attributes["end"] = this._end.ToString();
      }

      //Serialize children
      return rawNode;
    }

    public void Serialize(XmlElement element)
    {
        // Godot.GD.Print("Serializing to");
        var updatedRawNode = SerializeIntoRawNode();
        updatedRawNode.Serialize(element);
    }
    public System.String Get_zone_ref()
    {
      return this.zone_ref;
    }
    public void Set_zone_ref(System.String value)
    {
      this.zone_ref = value;
      this.NotifyChange();
    }
    public System.String Get_region_ref()
    {
      return this.region_ref;
    }
    public void Set_region_ref(System.String value)
    {
      this.region_ref = value;
      this.NotifyChange();
    }
    public System.String Get_side()
    {
      return this.side;
    }
    public void Set_side(System.String value)
    {
      this.side = value;
      this.NotifyChange();
    }
    public System.Int32 Get_start()
    {
      return this.start;
    }
    public void Set_start(System.Int32 value)
    {
      this.start = value;
      this.NotifyChange();
    }
    public System.Int32 Get_end()
    {
      return this.end;
    }
    public void Set_end(System.Int32 value)
    {
      this.end = value;
      this.NotifyChange();
    }


    public void DeserializeAtPath(string xpath, RawNode rawNode)
    {
      if(xpath.StartsWith("."))
      {
        xpath = xpath.Substring(1);
      }

      Deserialize(rawNode);
    }

    public void NotifyChange(List<ILinkedNode> linkedNodes)
    {
      if(_parentNode == null)
        return;
      linkedNodes.Add(this);
      _onSelfChangeCallbackList.ForEach(action => action(this));
      _onChangeCallbackList.ForEach(action => action(linkedNodes));
      _parentNode.NotifyChange(linkedNodes);
    }

    public void NotifyChange()
    {
      NotifyChange(new ());
    }

    public int? BuildIndexForChild(ILinkedNode linkedNode)
    {
      return null;
    }

    public bool IsValidChildType(ILinkedNode candidateChild) {
      return false;
    }

    public bool Equals(to? obj)
    {
        if (obj == null || GetType() != obj.GetType())
            return false;

        var other = (to)obj;
        return Equals(zone_ref, other.zone_ref) && Equals(region_ref, other.region_ref) && Equals(side, other.side) && Equals(start, other.start) && Equals(end, other.end);
    }

    public override int GetHashCode()
    {
        var acc = 0;

        acc = HashCode.Combine(acc, zone_ref);
        acc = HashCode.Combine(acc, region_ref);
        acc = HashCode.Combine(acc, side);
        acc = HashCode.Combine(acc, start);
        acc = HashCode.Combine(acc, end);
        return acc;
    }
  }
}

File: implementation/src/main/java/ro/anud/xml_xsd/implementation/model/WorldStep/Data/ZoneList/Zone/Region/Portals/Portal/To/To.java
package ro.anud.xml_xsd.implementation.model.WorldStep.Data.ZoneList.Zone.Region.Portals.Portal.To;
import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.*;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import ro.anud.xml_xsd.implementation.util.RawNode;

import java.util.*;
import ro.anud.xml_xsd.implementation.util.Subscription;
import java.util.function.Consumer;
import java.util.stream.Collectors;

import static ro.anud.xml_xsd.implementation.util.logging.LogScope.logScope;

  @EqualsAndHashCode
  @ToString
  @Builder
  @NoArgsConstructor
  @AllArgsConstructor
  @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
  public class To implements  ro.anud.xml_xsd.implementation.util.LinkedNode {

    public static String nodeName = "to";
    public static To fromRawNode(RawNode rawNode, ro.anud.xml_xsd.implementation.util.LinkedNode parent) {
      try (var logger = logScope()) {
        var instance = new To();
        if(Objects.nonNull(parent)) {
          instance.parentNode(parent);
        }
        instance.rawNode(rawNode);
        instance.deserialize(rawNode);
        return logger.logReturn(instance);
      }

    }
    public static To fromRawNode(RawNode rawNode) {
      try (var logger = logScope()) {
        var instance = fromRawNode(rawNode, null);
        return logger.logReturn(instance);
      }
    }
    public static Optional<To> fromRawNode(Optional<RawNode> rawNode, ro.anud.xml_xsd.implementation.util.LinkedNode parent) {
        try(var logger = logScope()) {
          return logger.logReturn(rawNode.map(o -> To.fromRawNode(o, parent)));
        }

    }
    public static List<To> fromRawNode(List<RawNode> rawNodeList, ro.anud.xml_xsd.implementation.util.LinkedNode parent) {
      try (var logger = logScope()) {
        List<To> returnList = Optional.ofNullable(rawNodeList)
            .orElse(List.of())
            .stream()
            .map(o -> To.fromRawNode(o, parent))
            .collect(Collectors.toList());
        return logger.logReturn(returnList);
      }
    }

    public String classTypeId() {
      return ".world_step.data.zone_list.zone.region.portals.portal.to";
    }

    //Attributes

    private String zoneRef;

    private String regionRef;

    private String side;

    private Integer start;

    private Integer end;

    //Children elements

    @ToString.Exclude()
    @EqualsAndHashCode.Exclude()
    @JsonIgnore
    @Builder.Default
    private RawNode rawNode = new RawNode();

    public RawNode rawNode() {
      return rawNode;
    }
    public void rawNode(RawNode rawNode) {
      this.rawNode = rawNode;
    }

    @ToString.Exclude()
    @EqualsAndHashCode.Exclude()
    @JsonIgnore
    @Builder.Default
    private Optional<ro.anud.xml_xsd.implementation.util.LinkedNode> parentNode = Optional.empty();

    public Optional<ro.anud.xml_xsd.implementation.util.LinkedNode> parentNode() {
      return parentNode;
    }

    @Builder.Default
    private List<Consumer<List<Object>>> onChangeList = new ArrayList<>();

    public String nodeName() {
      return "to";
    }

    public void notifyChange(List<Object> list) {
      try (var logger = logScope()) {
        list.addLast(this);
        logger.log("Notify change for", this.buildPath());
        onChangeList.forEach(consumer -> consumer.accept(list));
        parentNode.ifPresent(linkedNode -> linkedNode.notifyChange(list));
      }
    }

    public void parentNode(ro.anud.xml_xsd.implementation.util.LinkedNode linkedNode) {
      this.parentNode = Optional.of(linkedNode);
      notifyChange();
    }

    public void clearParentNode() {
      var parentNode = this.parentNode;
      this.parentNode = Optional.empty();
      parentNode.ifPresent(ro.anud.xml_xsd.implementation.util.LinkedNode::notifyChange);
    }

    public Optional<ro.anud.xml_xsd.implementation.model.WorldStep.Data.ZoneList.Zone.Region.Portals.Portal.Portal> parentAsPortal() {
      return parentNode.flatMap(node -> {
        if (node instanceof ro.anud.xml_xsd.implementation.model.WorldStep.Data.ZoneList.Zone.Region.Portals.Portal.Portal casted){
          return Optional.of(casted);
        }
        return Optional.empty();
      });
    }

    public void removeChild(Object object) {
    }

    public int buildIndexForChild(Object object) {
        return 0;
    }

    public void removeFromParent() {
      parentNode.ifPresent(node -> node.removeChild(this));
    }

    public Subscription onChange(Consumer<List<Object>> onChange) {
      try (var logger = logScope()) {
        onChangeList.add(onChange);
        return logger.logReturn(() -> onChangeList.remove(onChange));
      }
    }

    public void deserialize (RawNode rawNode) {
      try (var logger = logScope()) {
        this.rawNode = rawNode;
        var isDirty = false;
        try (var innerLogger = logScope("attributes")) {
          //Deserialize attributes
          innerLogger.log("zone_ref");
          var zoneRefValue = rawNode.getAttributeRequired("zone_ref");
          if(Objects.equals(this.zoneRef, zoneRefValue)) {
            isDirty = true;
          }
          this.zoneRef = zoneRefValue;
          innerLogger.log("region_ref");
          var regionRefValue = rawNode.getAttributeRequired("region_ref");
          if(Objects.equals(this.regionRef, regionRefValue)) {
            isDirty = true;
          }
          this.regionRef = regionRefValue;
          innerLogger.log("side");
          var sideValue = rawNode.getAttributeRequired("side");
          if(Objects.equals(this.side, sideValue)) {
            isDirty = true;
          }
          this.side = sideValue;
          innerLogger.log("start");
          var startValue = rawNode.getAttributeIntRequired("start");
          if(Objects.equals(this.start, startValue)) {
            isDirty = true;
          }
          this.start = startValue;
          innerLogger.log("end");
          var endValue = rawNode.getAttributeIntRequired("end");
          if(Objects.equals(this.end, endValue)) {
            isDirty = true;
          }
          this.end = endValue;
        }
        try (var innerLogger = logScope("children")) {
          //Deserialize children
        }

        if(isDirty) {
          notifyChange();
        }
      } catch (Exception e) {
        throw new RuntimeException("Deserialization failed for: " + this.buildPath(), e);
      }

    }

    public RawNode serializeIntoRawNode()
    {
      try (var logger = logScope()) {
        rawNode.setTag("to");
        try (var innerLogger = logScope("attributes")) {
          //Serialize attributes
          innerLogger.log("zone_ref");
          rawNode.setAttribute("zone_ref", this.zoneRef);
          innerLogger.log("region_ref");
          rawNode.setAttribute("region_ref", this.regionRef);
          innerLogger.log("side");
          rawNode.setAttribute("side", this.side);
          innerLogger.log("start");
          rawNode.setAttribute("start", this.start);
          innerLogger.log("end");
          rawNode.setAttribute("end", this.end);
        }
        try (var innerLogger = logScope("children")) {

          //Serialize children
          return rawNode;
        }
      }
    }

    public void serialize(Document document, Element element)
    {
        // Godot.GD.Print("Serializing to");
        var updatedRawNode = serializeIntoRawNode();
        updatedRawNode.populateNode(document, element);
    }

    public String getZoneRef()
    {
      return this.zoneRef;
    }
    public To setZoneRef(String value)
    {
      this.zoneRef = value;
      notifyChange();
      return this;
    }
    public String getRegionRef()
    {
      return this.regionRef;
    }
    public To setRegionRef(String value)
    {
      this.regionRef = value;
      notifyChange();
      return this;
    }
    public String getSide()
    {
      return this.side;
    }
    public To setSide(String value)
    {
      this.side = value;
      notifyChange();
      return this;
    }
    public Integer getStart()
    {
      return this.start;
    }
    public To setStart(Integer value)
    {
      this.start = value;
      notifyChange();
      return this;
    }
    public Integer getEnd()
    {
      return this.end;
    }
    public To setEnd(Integer value)
    {
      this.end = value;
      notifyChange();
      return this;
    }

    public ro.anud.xml_xsd.implementation.util.LinkedNode deserializeAtPath(String xpath, RawNode rawNode) {
       if(xpath.startsWith("."))
        {
          xpath = xpath.substring(1);
        }

        deserialize(rawNode);
        return this;
    }

    public Optional<ro.anud.xml_xsd.implementation.util.LinkedNode> getNodeAtPath(String xpath) {
       if(xpath.startsWith("."))
        {
          xpath = xpath.substring(1);
        }
        return Optional.of(this);
    }
  }


  /*
    dependant type:
    {
      "type": "element",
      "value": {
        "metaType": "object",
        "value": {},
        "isSingle": true,
        "isNullable": true,
        "attributes": {
          "metaType": "object",
          "value": {
            "zone_ref": {
              "metaType": "primitive",
              "value": "xs:string",
              "isNullable": false
            },
            "region_ref": {
              "metaType": "primitive",
              "value": "xs:string",
              "isNullable": false
            },
            "side": {
              "metaType": "primitive",
              "value": "type__rectangle_side",
              "isNullable": false
            },
            "start": {
              "metaType": "primitive",
              "value": "xs:int",
              "isNullable": false
            },
            "end": {
              "metaType": "primitive",
              "value": "xs:int",
              "isNullable": false
            }
          }
        }
      },
      "name": "to"
    }
  */